ui: Update nodes list with new badges

koorosh opened the pull request:
Resolves: #45478

Before, Badge component could accept string text as
input and it limited this component to provide more
specific content.

Now, this change allows to wrap Badge text with Tooltip
component and provide it as a text. It is necessary to add
tooltip on badges text and don't break styles and components
consistency.

Color codings for nodes and locality groups are different with
specific conditions that's why mapper functions were added to
map locality/node statuses to badges colors and badges text.

Release justification: bug fixes and low-risk updates to new functionality

<img width="1415" alt="Screenshot 2020-03-19 at 14 46 12" src="https://user-images.githubusercontent.com/3106437/77070675-573ff480-69f3-11ea-9cf8-7447bf0d9c40.png">

koorosh committed:
ui: Extend Badge component to accept arbitrary text

Before, Badge component could accept string text as
input and it limited this component to provide more
specific content.

Now, this change allows to wrap Badge text with Tooltip
component and provide it as a text.

Release note: None

Release justification: bug fixes and low-risk updates to new functionality

koorosh committed:
ui: Extend Badge component with all status styles

Before, Badge styles were defined only for few classes which
were used for that time. But now it is necessary to fill with
the rest of styles according to designs.

Release note: None

Release justification: bug fixes and low-risk updates to new functionality

koorosh committed:
ui: Update node list statuses with new Badge styles

Node list table on Cluster overview page has Status column
which contained only text representation of status.
But with new designs it has to be represented as Badges,
and color coded for better perception.

Release note: None

Release justification: bug fixes and low-risk updates to new functionality

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/45657)
<!-- Reviewable:end -->

dhartunian reviewed (changes requested):
This looks great!

One thing missing is badges in the Decommissioned Nodes box. Can you add those too?

> Reviewed 1 of 1 files at r1, 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45657#-:-M2svHi6A5xeGTtXFwFu:bsqkulw)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @koorosh)

___
*[pkg/ui/src/views/cluster/containers/nodesOverview/index.tsx, line 149 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/45657#-M2sv3wPB1h91tgtda5K:-M2sv3wPB1h91tgtda5L:b5q4pm6) ([raw file](https://github.com/cockroachdb/cockroach/blob/db06a803bf4111c2582c30e6a194d4fd9cad60cf/pkg/ui/src/views/cluster/containers/nodesOverview/index.tsx#L149)):*
> ```XML
>       return "danger";
>     default:
>       return undefined;
> ```

why not have a default badge type here? will this end up rendering as a "default" type anyway?
___
*[pkg/ui/src/views/cluster/containers/nodesOverview/index.tsx, line 281 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/45657#-M2su34P3HvPGsXPKF40:-M2su34P3HvPGsXPKF41:b-t6ueun) ([raw file](https://github.com/cockroachdb/cockroach/blob/db06a803bf4111c2582c30e6a194d4fd9cad60cf/pkg/ui/src/views/cluster/containers/nodesOverview/index.tsx#L281)):*
> ```XML
>         let status: string;
>         let tooltipText: string;
>         const badgeStatus = getBadgeTypeByNodeStatus(record.status);
> ```

This is difficult to read when `status` and `badgeStatus` are used together. What do you think of calling the variables `badgeText` and `badgeType`?


<!-- Sent from Reviewable.io -->

koorosh committed:
ui: Fix namings for provided Badge props on NodesOverview page

Rename variables to meaningful names to make it
clear what is their purpose.

Release note: None

Release justification: bug fixes and low-risk updates to new functionality

koorosh committed:
ui: Add switchExhaustiveCheck for static checking on default case

tslint rules don't allow to skip `default` case in switch
construction and in case of providing all possible case value
`default` value do not make a lot of sense.
Before, it required to provide 'undefined' or one of values.

Now, if `switchExhaustiveCheck` called with value allows to
statically ensure that all possible cases are covered.

Release note: None

Release justification: bug fixes and low-risk updates to new functionality

koorosh committed:
ui: Show decommissioned status as a Badge

Before, status on Decommissioned nodes table (Cluster Overview page) was
displayed as text.

Now, it is shown as <Badge> component with tooltip.

Release note: None

Release justification: bug fixes and low-risk updates to new functionality

koorosh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45657)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @koorosh)

___
*[pkg/ui/src/views/cluster/containers/nodesOverview/index.tsx, line 149 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/45657#-M2sv3wPB1h91tgtda5K:-M35io8AChQ8_uS7LRsL:be7swir) ([raw file](https://github.com/cockroachdb/cockroach/blob/db06a803bf4111c2582c30e6a194d4fd9cad60cf/pkg/ui/src/views/cluster/containers/nodesOverview/index.tsx#L149)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

why not have a default badge type here? will this end up rendering as a "default" type anyway?
</blockquote></details>

Yes, it will fall to "default" (as it is defined in defaultProps for <Badge> component)
Meanwhile, the main reason `default` case is added to avoid linting error for missed `default` case. To make it more clear, for `switch` construction where all possible cases are defined like this one - I added `switchExhaustiveCheck` utility function to statically check that `default` case won't be reached.

![Screenshot 2020-03-23 at 11.37.46.png](https://files.reviewableusercontent.io/images/reviewable/3106437/dJpCqIKholiMNygIJdZCkCPB/Screenshot%202020-03-23%20at%2011.37.46.png)
___
*[pkg/ui/src/views/cluster/containers/nodesOverview/index.tsx, line 281 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/45657#-M2su34P3HvPGsXPKF40:-M35kvqo8oBas3rmW1yf:bebrt1w) ([raw file](https://github.com/cockroachdb/cockroach/blob/db06a803bf4111c2582c30e6a194d4fd9cad60cf/pkg/ui/src/views/cluster/containers/nodesOverview/index.tsx#L281)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

This is difficult to read when `status` and `badgeStatus` are used together. What do you think of calling the variables `badgeText` and `badgeType`?
</blockquote></details>

Will do


<!-- Sent from Reviewable.io -->

dhartunian reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 2 of 2 files at r3, 1 of 1 files at r4, 1 of 1 files at r5, 1 of 1 files at r6, 2 of 2 files at r7, 1 of 1 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45657#-:-M3DWhQO3w3B3STb3tT1:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @dhartunian and @koorosh)



<!-- Sent from Reviewable.io -->

dhartunian commented:
bors r+

knz commented:
All these changes are user-facing and thus require a release note.
@dhartunian in the future, please also review release notes. I notice that many of the recent UI PRs did not get release notes.

@taroface our apologies - there's now a backlog of UI changes which you don't know about, but will need documenting. I encourage you to sync with @dhartunian to get a list, since you can't get it from the release note script.

taroface commented:
Thanks @knz - @dhartunian is there anything here that wouldn't have been reflected in my docs update at https://github.com/cockroachdb/docs/pull/6707 ?

dhartunian commented:
thanks for flagging this issue @knz. You're correct, I have not been good with enforcing release note hygiene. Will discuss with @taroface and @piyush-singh to make sure I have clarity here on how ui changes should get documented.
